web_poet/testing/pytest.py

```diff
@@ -0,0 +1,120 @@
+from web_poet.serialization import SerializedDataFileStorage, deserialize, load_class
+from web_poet.testing.utils import INPUT_DIR_NAME, META_FILE_NAME, OUTPUT_FILE_NAME
+from web_poet.utils import ensure_awaitable
+
+
+class WebPoetFile(pytest.File):
+    """Represents a directory containing test subdirectories for one Page Object."""
+
+    @staticmethod
+    def sorted(items: List["WebPoetItem"]) -> List["WebPoetItem"]:
+        """Sort the test list by the test name."""
+        return sorted(items, key=operator.attrgetter("name"))
+
+    def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:  # noqa: D102
+        result: List[WebPoetItem] = []
+        for entry in self.path.iterdir():
+            if entry.is_dir():
+                item: WebPoetItem = WebPoetItem.from_parent(
+                    self,
+                    name=entry.name,
+                    path=entry,
+                )
+                if item.input_path.is_dir() and item.output_path.is_file():
+                    result.append(item)
+        return self.sorted(result)
+
+
+class WebPoetItem(pytest.Item):
+    """Represents a directory containing one test."""
+
+    @property
+    def type_name(self) -> str:
+        """The name of the type being tested."""
+        if not self.parent:
+            raise ValueError("WebPoetItem has no parent")
+        return self.parent.name
+
+    @property
+    def input_path(self) -> Path:
+        """The inputs subdirectory path."""
+        return self.path / INPUT_DIR_NAME
+
+    @property
+    def output_path(self) -> Path:
+        """The output file path."""
+        return self.path / OUTPUT_FILE_NAME
+
+    @property
+    def meta_path(self) -> Path:
+        """The metadata file path."""
+        return self.path / META_FILE_NAME
+
+    def get_page(self) -> ItemPage:
+        """Return the page object created from the saved input."""
+        cls = load_class(self.type_name)
+        if not issubclass(cls, ItemPage):
+            raise TypeError(f"{self.type_name} is not a descendant of ItemPage")
+        storage = SerializedDataFileStorage(self.input_path)
+        return deserialize(cls, storage.read())
+
+    def get_meta(self) -> dict:
+        """Return the test metadata."""
+        if not self.meta_path.exists():
+            return {}
+        return json.loads(self.meta_path.read_bytes())
+
+    def get_output(self) -> dict:
+        """Return the output from the recreated Page Object."""
+        po = self.get_page()
+        item = asyncio.run(ensure_awaitable(po.to_item()))
+        return ItemAdapter(item).asdict()
+
+    def get_expected_output(self) -> dict:
+        """Return the saved output."""
+        return json.loads(self.output_path.read_bytes())
+
+    def runtest(self) -> None:  # noqa: D102
+        meta = self.get_meta()
+        frozen_time: Optional[str] = meta.get("frozen_time")
+        if frozen_time:
+            with freeze_time(frozen_time):
```

Comment:
Do you know how are timezones handled in freezegun? 

For example, on a machine with UTC+1 timezone frozen_time was recorded and stored to meta. Then, to run the test we load the time on another machine, which is in UTC+2 timezone. The code in the page object uses UTC time, or unix timestamps. Are these timestamps going to be different on those 2 machines?

Reply:
The string is paresed into a naive datetime object and then you can pass the timezone explicitly via https://github.com/spulec/freezegun#timezones

Without additional thinking (and recalling how do naive objects work) I'm not sure what this means for us, I'll check it.

Reply:
The datetime string, whatever it is, is converted to UTC if the timezone is present, or treated as UTC if it's not.

The code that runs under `freeze_time()` thinks that the current timezone is UTC, unless you pass a different timezone to `freeze_time()`, hopefully this shouldn't be a problem.

So when we generate the value we can either convert it to UTC or include the timezone, and for users who do that themselves we should say to do the same.

Reply:
Wouldn't assuming UTC by default be problematic?

So, let's say there is a Page Object:

```py
class MyPage(WebPage):    
    @field 
    def scraped_timestamp(self):
        # Unix timestamp. It doesn't depend on the TZ of the machine.
        return time.time()  

    @field 
    def scraped_datetime(self):
        # This one depends on the timezone.
        return str(datetime.datetime.now())
```

User runs generation code locally, and saves frozen_time without a timezone. MyPage assumes computer timezone when generating the item. But when freezegun is used, UTC is used instead. So, scraped_timestamp stays the same, but scraped_datetime returns a different value, and the comparison fails.

What do you think about requiring to set the timezone in the frozen_time, and not assuming UTC? Any other options?

I'm not sure converting frozen_time time to UTC before saving works. It seems that if we don't store timezone, and use either local TZ or UTC, then we can't guarantee scraped_datetime output would be the same.

Reply:
Yes, if the result depends on the timezone then we need to store both time and timezone. It looks like it will need to be a separate field, as `freeze_time` doesn't use the TZ specification from the time string to set a fake TZ and takes a separate argument for that. Alternatively we may be able to parse the string to extract the timezone in web-poet before passing it to `freeze_time`.

Reply:
I've tried to implement this but there are multiple open issues in freezegun related to its handling of timezones (I think I'm running into https://github.com/spulec/freezegun/issues/348 with my current code) and I worry that even if we manage to write code that works for our tests it can break on some user PO code. Also if the user generates the fixture manually and not with the scrapy-poet code (which can call freezegun in some complicated way if needed) we may need to give some instructions to them.

There is a newer alternative to freezegun, https://github.com/adamchainz/time-machine, I haven't tried it but it's a binary module. There may also be some workaround by e.g. requiring frozen_time to be in UTC (it shouldn't break user code, just reduce the cases that can be tested).

Reply:
So, our options:

1. Require that tests must be generated and run on machines with the same TZ. It looks pretty bad.
2. Work-around freezegun issues, or help fixing them. 
3. Handle timezones ourselves, e.g. using time.tzset(). Could be a simple solution. But could be not, we might start fighting against edge cases, increasing maintenance burden. Also, it doesn't work on Windows.
4. Try to use time-machine. Binary doesn't bother me too much. Cons: binary, doesn't work in Pypy.

> There may also be some workaround by e.g. requiring frozen_time to be in UTC (it shouldn't break user code, just reduce the cases that can be tested).

How would it work, could you please elaborate?

I also wonder if there is some other workaround possible, e.g. something about freezing the time before generating the output.

My votes:

1. If there is some creative work-around, let's do it. I think that's totally fine to be really particular about the format of frozen_time if this helps. For example, asking users to convert their local time to UTC is fine, as soon as scrapy-poet does the right thing. We may even go as far as not documenting the format of the time, and just saying it should be an output of some function which we provide.
2. Try time-machine. It quotes some other benefits, e.g. it says that freeze-gun can be super-slow in large projects, and we are going to have large projects.

Reply:
> I think that's totally fine to be really particular about the format of frozen_time if this helps. For example, asking users to convert their local time to UTC is fine, as soon as scrapy-poet does the right thing. 

The format is not the problem: as long as it's UTC or has the timezone information in it, the current code should return correct datetime values, they will just be in UTC, not in some other timezone. It's making the code run as if the timezone is different and still return correct values is problematic.

Reply:
Additional findings:

1. Both freezegun and time-machine freeze the fixture generation command if the asyncio reactor is used and time ticking is not enabled. It's enabled by default in time-machine and the doc says this: "code tends to make the (reasonable) assumption that time progresses whilst running, and should normally be tested as such. Testing with time frozen can make it easy to write complete assertions, but it's quite artificial. Write assertions against time ranges, rather than against exact values." This is a valid consideration but we currently compare exact values.
2. time-machine mocks the timezone using `time.tzset()` which doesn't work on Windows. It also requires `backports.zoneinfo` for this on Python < 3.8, but that shouldn't be a problem. This means that some tests will fail on Windows (ones for code that outputs data dependent on the current timezone) and we won't be able to ignore the affected fields unless we make some system that identifies such fields based on metadata or some logic. We can only recommend generating such tests using UTC.
3. I haven't finished the implementation but it looks possible to write all kinds of timezone-related tests with time-machine as opposed to freezegun and it won't require working around issues. It will require creating zoneinfo.ZoneInfo objects though so it will slightly complicate the code.
4. Technically, we don't need to use the same implementation in test generation and test running, but as scrapy-poet depends on web-poet it makes sense to use the same one.

Reply:
Do you think we can find some workaround for Windows users, even if it's ugly? Like asking hem to set timezone manually? :) Also, are date fields in zyte-common-items timezone-dependent?

Reply:
> Do you think we can find some workaround for Windows users, even if it's ugly? Like asking hem to set timezone manually?

Setting the timezone manually should work, no idea about lighter workarounds though.

> are date fields in zyte-common-items timezone-dependent?

`Metadata.dateDownloaded` is in UTC.